library/std/src/sys/pal/windows/fs.rs

```diff
@@ -346,6 +346,130 @@ impl File {
+                            // Wait failed
+                            Err(io::Error::last_os_error())
+                        } else {
+                            // WAIT_ABANDONED and WAIT_TIMEOUT should not be possible
+                            unreachable!()
+                        }
+                    } else {
+                        Err(err)
+                    }
+                }
+            };
+            c::CloseHandle(overlapped.hEvent);
+            final_result
+        }
+    }
+
+    pub fn lock(&self) -> io::Result<()> {
+        self.acquire_lock(c::LOCKFILE_EXCLUSIVE_LOCK)
+    }
+
+    pub fn lock_shared(&self) -> io::Result<()> {
+        self.acquire_lock(0)
+    }
+
+    pub fn try_lock(&self) -> io::Result<bool> {
+        let result = cvt(unsafe {
+            let mut overlapped = mem::zeroed();
+            c::LockFileEx(
+                self.handle.as_raw_handle(),
+                c::LOCKFILE_EXCLUSIVE_LOCK | c::LOCKFILE_FAIL_IMMEDIATELY,
+                0,
+                u32::MAX,
+                u32::MAX,
+                &mut overlapped,
+            )
+        });
+
+        if let Err(ref err) = result {
+            if err.raw_os_error() == Some(c::ERROR_IO_PENDING as i32)
+                || err.raw_os_error() == Some(c::ERROR_LOCK_VIOLATION as i32)
+            {
+                return Ok(false);
+            }
+        }
+        result?;
+        Ok(true)
+    }
+
+    pub fn try_lock_shared(&self) -> io::Result<bool> {
+        let result = cvt(unsafe {
+            let mut overlapped = mem::zeroed();
+            c::LockFileEx(
+                self.handle.as_raw_handle(),
+                c::LOCKFILE_FAIL_IMMEDIATELY,
+                0,
+                u32::MAX,
+                u32::MAX,
+                &mut overlapped,
+            )
+        });
+
+        if let Err(ref err) = result {
+            if err.raw_os_error() == Some(c::ERROR_IO_PENDING as i32)
+                || err.raw_os_error() == Some(c::ERROR_LOCK_VIOLATION as i32)
+            {
+                return Ok(false);
+            }
+        }
+        result?;
+        Ok(true)
+    }
+
+    pub fn unlock(&self) -> io::Result<()> {
+        // Unlock the handle twice because LockFileEx() allows a file handle to acquire
+        // both an exclusive and shared lock, in which case the documentation states that:
+        // "...two unlock operations are necessary to unlock the region; the first unlock operation
+        // unlocks the exclusive lock, the second unlock operation unlocks the shared lock"
+        cvt(unsafe { c::UnlockFile(self.handle.as_raw_handle(), 0, 0, u32::MAX, u32::MAX) })?;
+        let result =
+            cvt(unsafe { c::UnlockFile(self.handle.as_raw_handle(), 0, 0, u32::MAX, u32::MAX) });
+        if let Err(ref err) = result {
+            if err.raw_os_error() == Some(c::ERROR_NOT_LOCKED as i32) {
+                return Ok(());
+            }
+        }
+        result?;
+        Ok(())
```

Comment:
I'd combine this into a single `match` like so:
```suggestion
        match result {
            Ok(_) => Ok(()),
            Err(err) if err.raw_os_error() == Some(c::ERROR_NOT_LOCKED as i32) => Ok(()),
            Err(err) => Err(err),
        }
```

Reply:
Ah, I didn't know about that conditional match syntax. Thanks!